Compilation fails on Linux kernels >= 4.12

Due to kernel changes/cleanup, compilation of the rapiddisk kernel module fails with the following output:

```
/rapiddisk-dkms/usr/src/rapiddisk-5.0 $ make -C $(pwd) M=$PWD
make: Entering directory '/rapiddisk-dkms/usr/src/rapiddisk-5.0'
make -C/lib/modules/4.12.3-1-ARCH/build M=/rapiddisk-dkms/usr/src/rapiddisk-5.0
make[1]: Entering directory '/usr/lib/modules/4.12.3-1-ARCH/build'
  LD      /rapiddisk-dkms/usr/src/rapiddisk-5.0/built-in.o
  CC [M]  /rapiddisk-dkms/usr/src/rapiddisk-5.0/rapiddisk.o
/rapiddisk-dkms/usr/src/rapiddisk-5.0/rapiddisk.c: In function ‘attach_device’:
/rapiddisk-dkms/usr/src/rapiddisk-5.0/rapiddisk.c:708:26: error: ‘struct queue_limits’ has no member named ‘discard_zeroes_data’
  rdsk->rdsk_queue->limits.discard_zeroes_data = 1;
                          ^
make[2]: *** [scripts/Makefile.build:309: /rapiddisk-dkms/usr/src/rapiddisk-5.0/rapiddisk.o] Error 1
make[1]: *** [Makefile:1512: _module_/rapiddisk-dkms/usr/src/rapiddisk-5.0] Error 2
make[1]: Leaving directory '/usr/lib/modules/4.12.3-1-ARCH/build'
make: *** [Makefile:32: all] Error 2
make: Leaving directory '/rapiddisk-dkms/usr/src/rapiddisk-5.0'
```
According to this [mailing list entry](https://www.spinics.net/lists/raid/msg57009.html) the flag can completely be removed.

--- a/module/rapiddisk.c
+++ b/module/rapiddisk.c
@@ -705,7 +705,9 @@ static int attach_device(int size)
 	rdsk->rdsk_queue->nr_requests = nr_requests;
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,36)
 	rdsk->rdsk_queue->limits.discard_granularity = PAGE_SIZE;
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,12,0)
 	rdsk->rdsk_queue->limits.discard_zeroes_data = 1;
+#endif
 	rdsk->rdsk_queue->limits.max_discard_sectors = UINT_MAX;
 	queue_flag_set_unlocked(QUEUE_FLAG_DISCARD, rdsk->rdsk_queue);
 #endif

--- a/module/rapiddisk.c
+++ b/module/rapiddisk.c
@@ -508,6 +508,8 @@ rdsk_make_request(struct request_queue *q, struct bio *bio)
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,336)
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(4,8,0)
 	if (unlikely(bio_op(bio) == REQ_OP_DISCARD)) {
+#elif LINUX_VERSION_CODE >= KERNEL_VERSION(4,12,0)
+	if ((unlikely(bio_op(bio) == REQ_OP_DISCARD)) || (unlikely(bio_op(bio) == REQ_OP_WRITE_ZEROES))) {
 #else
 	if (unlikely(bio->bi_rw & REQ_DISCARD)) {
 #endif

--- a/src/archive.c
+++ b/src/archive.c
@@ -56,7 +56,7 @@ int archive_rd_volume(struct RD_PROFILE *prof, unsigned char *src,
 	in = (char *)malloc(BUFSZ);
 	out = (char *)malloc(BUFSZ);
 
-	if ((in == '\0') || (out == '\0')) {
+	if ((in == NULL) || (out == NULL)) {
 		printf("Error. Unable to acquire memory space.\n");
 		return -ENOMEM;
 	}
@@ -150,7 +150,7 @@ int restore_rd_volume(struct RD_PROFILE *prof, unsigned char *src,
 	in = (char *)malloc(BUFSZ);
 	out = (char *)malloc(BUFSZ);
 
-	if ((in == '\0') || (out == '\0')) {
+	if ((in == NULL) || (out == NULL)) {
 		printf("Error. Unable to acquire memory space.\n");
 		return -ENOMEM;
 	}

--- a/src/core.c
+++ b/src/core.c
@@ -352,7 +352,7 @@ int detach_device(struct RD_PROFILE *rd_prof, RC_PROFILE * rc_prof, unsigned cha
 		rc_prof = rc_prof->next;
 	}
 
-	if ((buf = (char *)malloc(BUFSZ)) == '\0') {
+	if ((buf = (char *)malloc(BUFSZ)) == NULL) {
 		printf("%s: malloc: Unable to allocate memory.\n", __func__);
 		return -1;
 	}
@@ -459,7 +459,7 @@ int cache_map(struct RD_PROFILE *rd_prof, struct RC_PROFILE * rc_prof,
 		rc_prof = rc_prof->next;
 	}
 
-	if ((buf = (char *)malloc(BUFSZ)) == '\0') {
+	if ((buf = (char *)malloc(BUFSZ)) == NULL) {
 		printf("%s: malloc: Unable to allocate memory.\n", __func__);
 		return -1;
 	}
@@ -547,7 +547,7 @@ int cache_unmap(struct RC_PROFILE *prof, unsigned char *string)
 		return -ENOENT;
 	}
 
-	if ((buf = (char *)malloc(BUFSZ)) == '\0') {
+	if ((buf = (char *)malloc(BUFSZ)) == NULL) {
 		printf("%s: malloc: Unable to allocate memory.\n", __func__);
 		return -ENOMEM;
 	}
@@ -599,7 +599,7 @@ int rdsk_flush(struct RD_PROFILE *rd_prof, RC_PROFILE *rc_prof, unsigned char *s
 		rc_prof = rc_prof->next;
 	}
 
-	if ((buf = (char *)malloc(BUFSZ)) == '\0') {
+	if ((buf = (char *)malloc(BUFSZ)) == NULL) {
 		printf("%s: malloc: Unable to allocate memory.\n", __func__);
 		return -ENOMEM;
 	}

